src/data/IANAData.tsx

```diff
@@ -124,27 +99,60 @@ export function addIANAVariantLocales(
   if (!variants) return;
 
   Object.values(variants).forEach((variant) => {
-    variant.languageCodes.forEach((prefix) => {
-      const bcpLang = languagesBySource.BCP[prefix];
-      if (!bcpLang) return;
-
-      const iso639_3 = bcpLang.ID;
-      const localeCode = `${prefix}_${variant.ID}`;
-
-      locales[localeCode] = {
-        ID: localeCode,
-        languageCode: iso639_3,
-        variantTagCode: variant.ID,
-        nameDisplay: variant.nameDisplay,
-        localeSource: 'IANA',
-        codeDisplay: localeCode,
-        type: ObjectType.Locale,
-        populationSource: PopulationSourceCategory.NoSource,
-        populationSpeaking: 0,
-        censusRecords: [],
-        territoryCode: '',
-        names: [variant.nameDisplay],
-      };
+    variant.localeCodes.forEach((fullCode) => {
+      
+      const parts = fullCode.split('-');
+      if (parts.length === 0) return;
+      const languageSubtag = parts[0];
+      const lang = languagesBySource.BCP[languageSubtag];
+      if (!lang) return;
+      const iso639_3 = lang.ID;
+
+      let scriptCode: string | undefined = undefined;
+      let regionCode = '';
+      const variantSubtags: string[] = [];
+
+      for (let i = 1; i < parts.length; i++) {
+        const subtag = parts[i];
+       
+        if (!scriptCode && /^[A-Z][a-z]{3}$/.test(subtag)) {
+          scriptCode = subtag;
+          continue;
+        }
+       
+        if (regionCode === '' && (/^[A-Z]{2}$/.test(subtag) || /^[0-9]{3}$/.test(subtag))) {
+          regionCode = subtag;
+          continue;
+        }
+       
+        variantSubtags.push(subtag);
+      }
+
+     
+      const idParts: string[] = [iso639_3];
+      if (scriptCode) idParts.push(scriptCode);
+      if (regionCode) idParts.push(regionCode);
+      if (variantSubtags.length > 0) idParts.push(...variantSubtags);
+      const localeCode = idParts.join('_');
```

Comment:
It's much cleaner if you use the logic we also used in `getLocaleCode`
```suggestion
      const localeCode = [iso639_3, scriptCode, regionCode, ...variantSubtags].filter(Boolean).join('_');
```